File: backtracking/46_permute/solution.py
from typing import List


def permute(nums):
  def recurse(current, output, num_set) -> List[List[int]]:
    if not num_set:
      output.append(current.copy())
      return output
    for num in nums:
      if num not in num_set: continue
      num_set.remove(num)
      current.append(num)
      recurse(current, output, num_set)
      current.pop()
      num_set.add(num)
    return output
  return recurse([], [], set(nums))

num_input = input("input some numbers like this '1,2,3' without quotation marks\n")
nums = num_input.split(",")
nums = list(map(int, nums))
result = permute(nums)
for arr in result: print(arr)

File: backtracking/46_permute/solution.js
function permute(nums) {
  function recurse(output, current, numSet) {
    if (numSet.size == 0) {
      output.push([...current]);
      return;
    }
    for (const num of nums) {
      if (!numSet.has(num)) continue;
      current.push(num);
      numSet.delete(num);
      recurse(output, current, numSet);
      current.pop();
      numSet.add(num);
    }
    return output;
  }
  return recurse([], [], new Set(nums));
}

if (process.argv.length !== 3) {
  console.warn("Usage: node solution.js int,int,...,int");
  process.exit(1);
}

input = process.argv[2].split(",").map(Number);

for (const n of input) {
  if (Number.isNaN(n)) {
    console.warn("Usage: node solution.js int,int,...,int");
    process.exit(1);
  }
}

console.log(permute(input));
